pkg/controller/argocd/keycloak.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2021 ArgoCD Operator Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package argocd
+
+import (
+	json "encoding/json"
+
+	"github.com/argoproj-labs/argocd-operator/pkg/common"
+	"github.com/argoproj-labs/argocd-operator/pkg/controller/argoutil"
+	appsv1 "github.com/openshift/api/apps/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	template "github.com/openshift/api/template/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+var (
+	privilegedMode bool  = false
+	graceTime      int64 = 75
+	portTLS        int32 = 8443
+)
+
+// getKeycloakContainerImage will return the container image for Keycloak.
+func getKeycloakContainerImage(img string, ver string) string {
+	return argoutil.CombineImageTag(img, ver)
+}
+
+func getKeycloakConfigMapTemplate(ns string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"description": "ConfigMap providing service ca bundle",
+				"service.beta.openshift.io/inject-cabundle": "true",
+			},
+			Labels: map[string]string{
+				"application": "${APPLICATION_NAME}",
+			},
+			Name:      "${APPLICATION_NAME}-service-ca",
+			Namespace: ns,
+		},
+		TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "ConfigMap"},
+	}
+}
+
+func getKeycloakContainer() corev1.Container {
+	return corev1.Container{
+		Env: []corev1.EnvVar{
+			{Name: "SSO_HOSTNAME", Value: "${SSO_HOSTNAME}"},
+			{Name: "DB_MIN_POOL_SIZE", Value: "${DB_MIN_POOL_SIZE}"},
+			{Name: "DB_MAX_POOL_SIZE", Value: "${DB_MAX_POOL_SIZE}"},
+			{Name: "DB_TX_ISOLATION", Value: "${DB_TX_ISOLATION}"},
+			{Name: "JGROUPS_PING_PROTOCOL", Value: "openshift.DNS_PING"},
+			{Name: "OPENSHIFT_DNS_PING_SERVICE_NAME", Value: "${APPLICATION_NAME}-ping"},
+			{Name: "OPENSHIFT_DNS_PING_SERVICE_PORT", Value: "8888"},
+			{Name: "X509_CA_BUNDLE", Value: "/var/run/configmaps/service-ca/service-ca.crt /var/run/secrets/kubernetes.io/serviceaccount/ca.crt"},
+			{Name: "JGROUPS_CLUSTER_PASSWORD", Value: "${JGROUPS_CLUSTER_PASSWORD}"},
+			{Name: "SSO_ADMIN_USERNAME", Value: "${SSO_ADMIN_USERNAME}"},
+			{Name: "SSO_ADMIN_PASSWORD", Value: "${SSO_ADMIN_PASSWORD}"},
```

Comment:
Cleartext passwords are viewable in the container's environment variables?

Reply:
yes the SSO_ADMIN_PASSWORD is viewable in the container's env variable.

oc exec sso-1-<> -- "env" | grep SSO_ADMIN_PASSWORD
---------------------------------------
SSO_ADMIN_PASSWORD=cTWjSwxS